system/include/libc/alltypes.h.in

```diff
@@ -50,7 +50,7 @@ TYPEDEF unsigned long pthread_t;
 TYPEDEF struct __pthread * pthread_t;
 #endif
 TYPEDEF int pthread_once_t;
-TYPEDEF unsigned pthread_key_t;
+TYPEDEF unsigned long pthread_key_t;
```

Comment:
I thought we decided this wasn't needed?

Reply:
I remember we talked about implementing these functions in a different way, can you remind me how? 
https://github.com/emscripten-core/emscripten/blob/c92bbd565a8c0732dbd1fc2dfccd78a3c58a89e3/system/lib/pthread/library_pthread_stub.c#L301-L320

Reply:
Hmm... how about this:

```
static size_t num_tls_entries = 0;
static void** tls_entries;
static bool* tls_entries_occupied;

int pthread_key_create() {
   // find empty slot in tls_entries.. realloc if needed
   // O(N) scan should be fine here
   return entry_index;
}

int pthread_key_delete() {
   // check entry N is valid.
   // mark entry N as unused
}

void* pthread_getspecific() {
   // fail if entry is unused
   return entry[N];
}
```

Its not the most efficient `pthread_key_create` but I think it should be fine given that these are just stubs for single threaded apps .. basically fakes.  Any app who has a bottleneck here would be way better off building a real thread-less build (i.e. ifdefing that threads out of their code completely).

Reply:
And why is changing the type from `unsigned` to `unsigned long` not better? it doesn't affect `wasm32`, only `wasm64`.

Or why not just return `NULL` from those functions? Almost all functions in that file return `0` and do nothing.. I am not sure how any single threaded code is going to be disappointed with these functions failing.

I don't mind adding the code you suggested, but it is actually not that minimal, reallocating and copying 2 arrays, scanning them etc.. that is a fair bit of code I'd rather not add.

Why `tls_entries_occupied` btw? Can't entries in `tls_entries` simply be `NULL` ?

I wonder how much C code out there is effectively implementing `std::vector` over and over.

Reply:
The main reason not to change is to avoid differing from upstream musl I think. 

I'm not sure how to weight the cost of that against the cost of this implementation.     I wonder if we could find a existing "fake-single-threaded-pthread-impl" that we could drop in here maybe?

I'm not sure the history of this file or why the original author found it worthwhile to fake pthread TLS in particular.

Reply:
ok, I will add the code then..